### PR TITLE
docs: document the Plugin and DynPlugin trait

### LINKED ISSUE #791: Document the Plugin and DynPlugin trait in our code

### PR DESCRIPTION
**Pull request by bnjjj**
close #791

### CODE REVIEW

File: apollo-router-core/src/plugin.rs [hunk 1]
```diff
@@ -64,41 +64,59 @@ pub fn plugins() -> HashMap<String, PluginFactory> {
     PLUGIN_REGISTRY.lock().expect("Lock poisoned").clone()
 }
 
+/// All router plugins must implement the Plugin trait. This trait defines lifecycle hooks that enable hooking into Apollo Router services.
+/// The trait also provides a default implementations for each hook, which returns the associated service unmodified.
```
Comment (garypen) [inline apollo-router-core/src/plugin.rs:68]: Add a link to the online docs: https://www.apollographql.com/docs/router/customizations/native/#plugin-lifecycle

File: apollo-router-core/src/plugin.rs [hunk 2]
```diff
@@ -64,41 +64,59 @@ pub fn plugins() -> HashMap<String, PluginFactory> {
...
+    /// Define `execution_service` if your customization includes logic to govern execution (for example, if you want to block a particular query based on a policy decision).
     fn execution_service(
         &mut self,
         service: BoxService<ExecutionRequest, ExecutionResponse, BoxError>,
     ) -> BoxService<ExecutionRequest, ExecutionResponse, BoxError> {
         service
     }
 
+    /// This service handles communication between the Apollo Router and your subgraphs.
+    /// Define `subgraph_service` to configure this communication (for example, to dynamically add headers to pass to a subgraph).
```
Comment (garypen) [inline apollo-router-core/src/plugin.rs:119]: Perhaps note that "_name" is the name of the subgraph service and can be used to customize processing by subgraph if desired? (same for DynPlugin)

**Review by garypen — CHANGES_REQUESTED** [review]
Looks great! I think the link to online docs will help provide more context about lifecycle behaviour.

**Review by garypen — APPROVED** [review]

File: apollo-router-core/src/plugin.rs [hunk 3]
```diff
@@ -64,49 +64,71 @@ pub fn plugins() -> HashMap<String, PluginFactory> {
...
         Ok(())
     }
+
+    /// This is invoked whenever configuration is changed
+    /// during router execution, including at shutdown.
     async fn shutdown(&mut self) -> Result<(), BoxError> {
         Ok(())
     }
 
+    /// This service runs at the very beginning and very end of the request lifecycle.
```
Comment (Geal) [inline apollo-router-core/src/plugin.rs:91]: maybe beginning and end is not clear here, it should emphasize that it runs when receiving the request and before sending the response

**Review by Geal — APPROVED** [review]

File: apollo-router-core/src/plugin.rs (hunk 3 above)
Comment (bnjjj) [reply apollo-router-core/src/plugin.rs:91]: Added `It's the entrypoint of every requests and also the last hook before sending the response.`

**Pull request merged**